### ISSUE TITLE
Requirement failed: size mismatch

### ISSUE THREAD

**Issue by m-developer96** [intel-analytics/BigDL#3039]
Hi, I used the TextClassification example and create my own model with these configurations:

```
# it said that batch_size must be dividable by core_num and my core_num is 6
batch_size = 66 
embedding_dim = 300
sequence_len = 32
max_words = 20000
training_split = 0.8
```
And my model :

```
model_type = 'lstm'
p = 0.2

def build_model(class_num):
    model = Sequential()
    if model_type.lower() == "cnn":
        model.add(Reshape([embedding_dim, 1, sequence_len]))
        model.add(SpatialConvolution(embedding_dim, 128, 5, 1))
        model.add(ReLU())
        model.add(SpatialMaxPooling(5, 1, 5, 1))
        model.add(SpatialConvolution(128, 128, 5, 1))
        model.add(ReLU())
        model.add(SpatialMaxPooling(5, 1, 5, 1))
        model.add(Reshape([128]))
    elif model_type.lower() == "lstm":
        model.add(Recurrent()
                  .add(LSTM(embedding_dim, 128, p)))
        model.add(Select(2, -1))
    elif model_type.lower() == "gru":
        model.add(Recurrent()
                  .add(GRU(embedding_dim, 128)))
        model.add(Select(2, -1))
    else:
        raise ValueError('model can only be cnn, lstm, or gru')
       
    model.add(Reshape([128]))
    model.add(Linear(128, 100))\
        .add(Dropout(0.3))\
        .add(LeakyReLU(negval=0.3,inplace=False,bigdl_type="float"))\
        .add(Linear(100, class_num))\
        .add(LeakyReLU(negval=0.3,inplace=False,bigdl_type="float"))\
        .add(SoftMax())   
    
    return model
```

And my optimizer:

```
    optimizer = Optimizer(
        model=model,
        training_rdd=train_rdd,
        criterion=BCECriterion(),
        end_trigger=MaxEpoch(max_epoch),
        batch_size=batch_size,
        optim_method= Adam(learningrate=lr))

    optimizer.set_validation(
        batch_size=batch_size,
        val_rdd=val_rdd,
        trigger=EveryEpoch(),
        val_method=[Top1Accuracy(),Loss()]
    )
```

But when I run `optimizer.optimize()`, I got this error:

```
Layer info: Sequential[f511c363]{
  [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> (8) -> output]
  (1): Recurrent[c7ca6fb5]ArrayBuffer(LSTM(300, 128, 0.2))
  (2): nn.Select
  (3): Reshape[4c601722](128)
  (4): Linear[8f117063](128 -> 100)
  (5): Dropout[fd5f9514](0.3)
  (6): LeakyReLU[cc1306d7]
  (7): Linear[9a86e28e](100 -> 2)
  (8): SoftMax[31ec26cb]
}/Recurrent[c7ca6fb5]ArrayBuffer(LSTM(300, 128, 0.2))/LSTM(300, 128, 0.2)/Sequential[f5d337dc]{
  [input -> (1) -> (2) -> (3) -> output]
  (1): nn.Flatten
  (2): StaticGraph[82824721]
  (3): ConcatTable[f86d629d] {
	input
	  |`-> (1): SelectTable(1)
	   `-> (2): NarrowTable[e6059f37](2, 2)
	   ... -> output
}
}/StaticGraph[82824721]/Linear[4da41ee9](300 -> 128)
java.lang.IllegalArgumentException: requirement failed: size mismatch, m1:11x32 m2:300x128
```

I couldn't know how should I check my input or model to fix this error. Could anyone help me, please?

With many thanks:)

**Comment by m-developer96** [comment]
I think m1:11x32 related to each core job (66/6=11) and my seq_len = 32, but I don't know how does it act here!

**Comment by hkvision** [comment]
Hi @m-developer96 

Yeah, 11 is the batch size for one core.

I suppose you have something wrong with your input. You need to perform embedding on your input beforehand. To be more specific, the input to LSTM should be `(batch, seq_len, embed_dim)` and in your case `(batch, 32, 300)`, I suppose your input is only `(batch, 32)` and therefore the batch dimension is mistreated as the seq_len and the seq_len is mistreated as the embed_dim. Please correct your input and have another try.

**Comment by hkvision** [comment]
As for your reference, you are recommended to run and understand our text classification example first before you modifying to your own model or using your own data. This would prevent you many of unexpected errors.

**Comment by jason-dai** [comment]
@m-developer96 It is probably easier to use the Keras API to define your model; see examples at https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/examples/textclassification/text_classification.py and https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/models/textclassification/text_classifier.py

**Comment by m-developer96** [comment]
Hi @hkvision 
I run [this example](https://github.com/intel-analytics/BigDL/blob/branch-0.1/pyspark/example/tutorial/simple_text_classification/text_classfication.ipynb) completely. Then I want to create my own model and I just changed batch_size, seq_len, and embedding_dim.
As you said, I checked my input. This is the first row of my train_rdd:
`Sample: features: [JTensor: storage: [ 0.012  -0.017   0.0228 ...  0.      0.      0.    ], shape: [300  31], float], labels: [JTensor: storage: [2.], shape: [1], float]`

Where is the problem ?

**Comment by m-developer96** [comment]
@jason-dai 
Thanks, I'll try it soon.

**Comment by hkvision** [comment]
> Hi @hkvision
> I run [this example](https://github.com/intel-analytics/BigDL/blob/branch-0.1/pyspark/example/tutorial/simple_text_classification/text_classfication.ipynb) completely. Then I want to create my own model and I just changed batch_size, seq_len, and embedding_dim.
> As you said, I checked my input. This is the first row of my train_rdd:
> `Sample: features: [JTensor: storage: [ 0.012 -0.017 0.0228 ... 0. 0. 0. ], shape: [300 31], float], labels: [JTensor: storage: [2.], shape: [1], float]`
> 
> Could you guide me plz? Where is the problem?

Hi @m-developer96 

The shape of your tensor is wrong, it should be (seq_len, embed_dim), namely `[32, 300]` in your case, why is `[300, 31]`. I suppose there's something wrong with your preprocessing. Please make sure you have applied the embeddings to each token correctly.

**Comment by m-developer96** [comment]
@hkvision
It's because of `transpose(1, 0)` that written in `to_sample()` method. I remove it and tried before you but in this case, I got this error:

`Requirement failed: input size should be equal to target size, but got input size: List(10, 2), target size: List(10, 1)`

But in my model. last layer is `Linear(100, 2)` and a `Softmax()` layer after that. Basically, the model has 2 output, is that correct?
Then why I got that error?

```

def build_model(class_num):
    model = Sequential()
    if model_type.lower() == "cnn":
        model.add(Reshape([embedding_dim, 1, sequence_len]))
        model.add(SpatialConvolution(embedding_dim, 128, 5, 1))
        model.add(ReLU())
        model.add(SpatialMaxPooling(5, 1, 5, 1))
        model.add(SpatialConvolution(128, 128, 5, 1))
        model.add(ReLU())
        model.add(SpatialMaxPooling(5, 1, 5, 1))
        model.add(Reshape([128]))
    elif model_type.lower() == "lstm":
        model.add(Recurrent()
                  .add(LSTM(embedding_dim, 128, p)))
        model.add(Select(2, -1))
    elif model_type.lower() == "gru":
        model.add(Recurrent()
                  .add(GRU(embedding_dim, 128)))
        model.add(Select(2, -1))
    else:
        raise ValueError('model can only be cnn, lstm, or gru')
    
    
    model.add(Reshape([128]))
    model.add(Linear(128, 100))\
        .add(Dropout(0.3))\
        .add(LeakyReLU(negval=0.3,inplace=False,bigdl_type="float"))\
        .add(Linear(100, class_num))\
        .add(SoftMax())   
```

**Comment by m-developer96** [comment]
It's because of `transpose(1, 0)` that written in `to_sample()` method.
Thanks :)

**Comment by hkvision** [comment]
No problem! Glad to hear that the issue gets resolved!
